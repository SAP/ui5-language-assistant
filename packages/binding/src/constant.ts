export const BINDING_ISSUE_TYPE = "binding-issue";
export const PROPERTY_BINDING_INFO =
  "sap.ui.base.ManagedObject.PropertyBindingInfo";
export const AGGREGATION_BINDING_INFO =
  "sap.ui.base.ManagedObject.AggregationBindingInfo";
export const Binding_Mode = "sap.ui.model.BindingMode";
