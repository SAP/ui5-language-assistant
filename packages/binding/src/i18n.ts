import type { i18n, TFunction } from "i18next";
import { DEFAULT_I18N_NAMESPACE } from "@ui5-language-assistant/context";
import i18nEn from "./i18n/i18n.json";
import i18next from "i18next";

export function initI18n(i18nInstance: i18n): void {
  i18nInstance.addResourceBundle(
    i18nInstance.language,
    DEFAULT_I18N_NAMESPACE,
    i18nEn
  );
}

export const adaptTranslatedText = (text: string): string => {
  if (/&quot;/.test(text)) {
    text = text.replace(/&quot;/g, '"');
  }
  if (/&#39;/.test(text)) {
    text = text.replace(/&#39;/g, "'");
  }
  if (/&amp;quot;/.test(text)) {
    text = text.replace(/&amp;quot;/g, "&quot;");
  }
  return text;
};

export const t: TFunction = (key: string, ...args) => {
  const result = i18next.t(key, ...args);
  return adaptTranslatedText(result);
};
