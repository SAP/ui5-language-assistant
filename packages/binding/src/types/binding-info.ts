import { Context } from "@ui5-language-assistant/context";
import {
  MarkupKind,
  TextDocumentPositionParams,
} from "vscode-languageserver-protocol";

export enum TypeKind {
  integer = "integer",
  Integer = "integer",
  string = "string",
  String = "string",
  function = "string",
  Function = "string",
  boolean = "boolean",
  Boolean = "boolean",
  object = "object",
  Object = "object",
  PropertyBindingInfo = "object",
  any = "any",
}
export enum BindingInfoName {
  // common between PropertyBindingInfo and AggregationBindingInfo
  events = "events",
  model = "model",
  path = "path",
  parameters = "parameters",
  suspended = "suspended",
  // PropertyBindingInfo only
  constraints = "constraints",
  formatter = "formatter",
  formatOptions = "formatOptions",
  mode = "mode",
  parts = "parts",
  type = "type",
  targetType = "targetType",
  useRawValues = "useRawValues",
  useInternalValues = "useInternalValues",
  value = "value",
  // AggregationBindingInfo only
  factory = "factory",
  filters = "filters",
  groupHeaderFactory = "groupHeaderFactory",
  key = "key",
  length = "length",
  sorter = "sorter",
  startIndex = "startIndex",
  template = "template",
  templateShareable = "templateShareable",
}
export interface Dependents {
  name: string;
  type: PropertyType[];
}
/**
 * Represents the type of a property.
 */
export interface PropertyType {
  /**
   * The kind of the property type.
   */
  kind: TypeKind;

  /**
   * Specifies the possible values for the property.
   */
  possibleValue?: {
    /**
     * An array of possible values for the property.
     */
    values: (string | boolean)[];

    /**
     * Indicates whether the possible values are fixed or not.
     */
    fixed: boolean;
  };

  /**
   * An array of dependent properties.
   */
  dependents: Dependents[];

  /**
   * An array of names of elements that are not allowed for this property.
   */
  notAllowedElements: string[];

  /**
   * Specifies the possible elements for this property.
   */
  possibleElements?: BindingInfoElement[];

  /**
   * Indicates whether the property is a collection or not.
   */
  collection?: boolean;
  /**
   * Indicates reference to other element
   */
  reference?: string;
}

export interface BindingInfoElement {
  name: string;
  type: PropertyType[];
  required?: boolean;
  documentation: {
    kind: MarkupKind;
    value: string;
  };
}

export interface BindContext extends Context {
  textDocumentPosition?: TextDocumentPositionParams;
  doubleQuotes?: boolean;
}

export enum ClassName {
  Sorter = "Sorter",
  Filter = "Filter",
}

export enum Operator {
  All = "All",
  Any = "Any",
  BT = "BT",
  Contains = "Contains",
  EndsWith = "EndsWith",
  EQ = "EQ",
  GE = "GE",
  GT = "GT",
  LE = "LE",
  LT = "LT",
  NB = "NB",
  NE = "NE",
  NotContains = "NotContains",
  NotEndsWith = "NotEndsWith",
  NotStartsWith = "NotStartsWith",
  StartsWith = "StartsWith",
}

export enum BindingMode {
  "sap.ui.model.BindingMode.Default" = "sap.ui.model.BindingMode.Default",
  "sap.ui.model.BindingMode.OneTime" = "sap.ui.model.BindingMode.OneTime",
  "sap.ui.model.BindingMode.OneWay" = "sap.ui.model.BindingMode.OneWay",
  "sap.ui.model.BindingMode.TwoWay" = "sap.ui.model.BindingMode.TwoWay",
}
