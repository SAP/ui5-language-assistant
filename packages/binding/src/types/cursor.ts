import { BindingParserTypes as BindingTypes } from "@ui5-language-assistant/binding-parser";
export interface BaseContext {
  element: BindingTypes.StructureElement;
}
/**
 * Initial scenario
 * a. ""
 */
export interface InitialContext {
  type: "initial";
  kind: "expression-binding"; // provides expression binding as code completion item
}
/**
 * Empty scenario
 * a. { }
 */
export interface EmptyContext {
  type: "empty";
  kind: "properties"; // all properties
}
/**
 * Key context scenarios
 *
 * a. `<CURSOR>`keyProperty
 *
 * b. keyProperty`<CURSOR>`
 *
 * c. key`<CURSOR>`Property
 */
export interface KeyContext extends BaseContext {
  type: "key";
  kind: "properties-excluding-duplicate"; // all properties excluding already used properties
}

/**
 * Value context scenarios
 *
 * a. keyProperty: `<CURSOR>`
 *
 * b. keyProperty: `<CURSOR>`'value-for-this-key'
 *
 * c. keyProperty: `<CURSOR>`  'value-for-this-key' [space]
 *
 * d. keyProperty: 'value-for`<CURSOR>`-this-key'
 *
 * e. keyProperty: 'value-for-this-key'`<CURSOR>`
 *
 * g. keyProperty `<CURSOR>` 'value-for-this-key' [missing colon]
 *
 * h. keyProperty `<CURSOR>` [space(s)] [missing colon]
 */
export interface ValueContext extends BaseContext {
  type: "value";
  kind: "value";
}
/**
 * Key value context scenarios
 *
 * a. keyProperty: 'value-for-this-key'  `<CURSOR>` [spaces]
 *
 * b. keyProperty: 'value-for-this-key', `<CURSOR>` [comma]
 *
 * c. `<CURSOR>` keyProperty: 'value-for-this-key'
 *
 * d. keyProperty: 'value-for-this-key',`<CURSOR>`, [between comma]
 */
export interface KeyValueContext {
  type: "key-value";
  kind: "properties-with-value-excluding-duplicate";
}
export type CursorContext =
  | InitialContext
  | EmptyContext
  | KeyContext
  | ValueContext
  | KeyValueContext;
