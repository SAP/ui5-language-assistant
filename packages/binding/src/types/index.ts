export type { BindingIssue } from "./issues";
export type {
  PropertyBindingInfoElement,
  Dependents,
  PropertyType,
  BindContext,
} from "./binding-info";
export { BindingInfoName, TypeKind } from "./binding-info";

export { BINDING_ISSUE_TYPE } from "../../src/constant";

export type {
  CursorContext,
  KeyValueContext,
  BaseContext,
  EmptyContext,
  InitialContext,
  KeyContext,
  ValueContext,
} from "./cursor";
