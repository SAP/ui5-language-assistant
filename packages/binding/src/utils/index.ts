import type {
  UI5Aggregation,
  UI5Type,
} from "@ui5-language-assistant/semantic-model-types";

export {
  typesToValue,
  valueTypeMap,
  isParts,
  isAnyType,
  findRange,
  defaultRange,
  getPropertyTypeWithPossibleValue,
  isMacrosMetaContextPath,
} from "./element";

export { getCursorContext } from "./cursor";

export { getLogger } from "./logger";

export { getDocumentation } from "./documentation";

/**
 * Finds and returns the first alternative type in the aggregation's altTypes array
 * that has a kind property equal to "PrimitiveType".
 *
 * @param {UI5Aggregation} [aggregation] - The aggregation object which may contain alternative types.
 * @returns {UI5Type | undefined} - The first alternative type with kind "PrimitiveType", or undefined if not found.
 */
export const findPrimitiveTypeInAggregation = (
  aggregation?: UI5Aggregation
): UI5Type | undefined =>
  aggregation?.altTypes?.find((i) => i.kind === "PrimitiveType");

export {
  getReference,
  buildType,
  getPossibleValuesForClass,
  getConstructorParameterProperties,
  getPossibleElement,
} from "./definition";
