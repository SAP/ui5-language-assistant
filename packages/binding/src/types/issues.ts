import { XMLViewIssueSeverity } from "@ui5-language-assistant/xml-views-validation";
import { Range } from "vscode-languageserver-types";
import { BINDING_ISSUE_TYPE } from "../constant";

export type BindingIssue =
  | MissingKey
  | MissingColon
  | TooManyColons
  | MissingValue
  | MissingComma
  | TooManyCommas
  | Syntax
  | TrailingComma
  | UnknownChar
  | UnknownPropertyBindingInfo
  | MissMatchValue
  | NotAllowedProperty
  | DuplicateProperty
  | RequiredDependency
  | RecursiveProperty
  | Unnecessary
  | MissingBracket;

interface BaseUI5XMLViewBindingIssue {
  issueType: typeof BINDING_ISSUE_TYPE;
  range: Range;
  kind: string;
  message: string;
  severity: XMLViewIssueSeverity;
}

export interface MissingKey extends BaseUI5XMLViewBindingIssue {
  kind: "MissingKey";
}
export interface MissingColon extends BaseUI5XMLViewBindingIssue {
  kind: "MissingColon";
}
export interface TooManyColons extends BaseUI5XMLViewBindingIssue {
  kind: "TooManyColons";
}
export interface MissingValue extends BaseUI5XMLViewBindingIssue {
  kind: "MissingValue";
}
export interface MissingComma extends BaseUI5XMLViewBindingIssue {
  kind: "MissingComma";
}
export interface TooManyCommas extends BaseUI5XMLViewBindingIssue {
  kind: "TooManyCommas";
}
export interface TrailingComma extends BaseUI5XMLViewBindingIssue {
  kind: "TrailingComma";
}
export interface MissingBracket extends BaseUI5XMLViewBindingIssue {
  kind: "MissingBracket";
}
/**
 * All parser error
 */
export interface Syntax extends BaseUI5XMLViewBindingIssue {
  kind: "Syntax";
}
/**
 * All unknown character
 */
export interface UnknownChar extends BaseUI5XMLViewBindingIssue {
  kind: "UnknownChar";
}

export interface UnknownPropertyBindingInfo extends BaseUI5XMLViewBindingIssue {
  kind: "UnknownPropertyBindingInfo";
}
export interface MissMatchValue extends BaseUI5XMLViewBindingIssue {
  kind: "MissMatchValue";
}
export interface NotAllowedProperty extends BaseUI5XMLViewBindingIssue {
  kind: "NotAllowedProperty";
}
export interface DuplicateProperty extends BaseUI5XMLViewBindingIssue {
  kind: "DuplicateProperty";
}
export interface RequiredDependency extends BaseUI5XMLViewBindingIssue {
  kind: "RequiredDependency";
}
export interface Unnecessary extends BaseUI5XMLViewBindingIssue {
  kind: "Unnecessary";
}
export interface RecursiveProperty extends BaseUI5XMLViewBindingIssue {
  kind: "RecursiveProperty";
}
