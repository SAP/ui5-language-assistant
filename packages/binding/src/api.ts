import type { BindingIssue } from "./types";
import { BINDING_ISSUE_TYPE } from "./types";

export { getCompletionItems } from "./services/completion";
export { bindingValidators } from "./services/diagnostics";
export { getHover } from "./services/hover";
export type { BindingIssue } from "./types";

export function isBindingIssue<T extends { issueType: string }>(
  issue: BindingIssue | T
): issue is BindingIssue {
  return issue.issueType === BINDING_ISSUE_TYPE;
}

export {
  AGGREGATION_BINDING_INFO,
  PROPERTY_BINDING_INFO,
  FILTER_OPERATOR,
} from "./constant";

export { getBindingElements } from "./definition/definition";
