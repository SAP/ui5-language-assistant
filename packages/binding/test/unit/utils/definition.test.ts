import {
  UI5Class,
  UI5ConstructorParameters,
  UI5Type,
} from "@ui5-language-assistant/semantic-model-types";
import {
  getReference,
  getPossibleValuesForClass,
  buildType,
  getConstructorParameterProperties,
  getPossibleElement,
} from "../../../src/utils";
import { BindContext } from "../../../src/types";
import { OPEN_FRAMEWORK } from "@ui5-language-assistant/constant";

function createType<T>(param: unknown): T {
  return param as T;
}

describe("getReference", () => {
  it('should return "filters" for PrimitiveType with name Filter', () => {
    const type = { kind: "PrimitiveType", name: "Filter" };
    expect(getReference(createType<UI5Type>(type))).toBe("filters");
  });

  it('should return "filters" for UI5Enum with name Filter', () => {
    const type = { kind: "UI5Enum", name: "Filter" };
    expect(getReference(createType<UI5Type>(type))).toBe("filters");
  });

  it('should return "filters" for UI5Class with name Filter', () => {
    const type = { kind: "UI5Class", name: "Filter" };
    expect(getReference(createType<UI5Type>(type))).toBe("filters");
  });

  it('should return "filters" for UI5Typedef with name Filter', () => {
    const type = { kind: "UI5Typedef", name: "Filter" };
    expect(getReference(createType<UI5Type>(type))).toBe("filters");
  });

  it("should return undefined for PrimitiveType with different name", () => {
    const type = { kind: "PrimitiveType", name: "Other" };
    expect(getReference(createType<UI5Type>(type))).toBeUndefined();
  });

  it("should return reference for ArrayType", () => {
    const type = {
      kind: "ArrayType",
      type: { kind: "PrimitiveType", name: "Filter" },
    };
    expect(getReference(createType<UI5Type>(type))).toBe("filters");
  });

  it("should return reference for UnionType", () => {
    const type = {
      kind: "UnionType",
      types: [
        { kind: "PrimitiveType", name: "Other" },
        { kind: "PrimitiveType", name: "Filter" },
      ],
    };
    expect(getReference(createType<UI5Type>(type))).toBe("filters");
  });

  it("should return undefined for unknown type", () => {
    const type = { kind: "UnknownType", name: "Filter" };
    expect(getReference(createType<UI5Type>(type))).toBeUndefined();
  });
});

describe("getPossibleValuesForClass", () => {
  it("should return possible values for a given UI5 class", () => {
    const context = {
      ui5Model: {
        classes: {
          ClassA: { extends: { name: "BaseClass", kind: "class" } },
          ClassB: { extends: { name: "BaseClass", kind: "class" } },
          ClassC: { extends: { name: "OtherClass", kind: "class" } },
        },
      },
    };

    const type = { name: "BaseClass", kind: "class" };

    const result = getPossibleValuesForClass(
      createType<BindContext>(context),
      createType<UI5Class>(type)
    );

    expect(result).toEqual(["ClassA", "ClassB"]);
  });

  it("should return an empty array if no classes extend the given UI5 class", () => {
    const context = {
      ui5Model: {
        classes: {
          ClassA: { extends: { name: "OtherClass", kind: "class" } },
          ClassB: { extends: { name: "OtherClass", kind: "class" } },
        },
      },
    };

    const type = { name: "BaseClass", kind: "class" };

    const result = getPossibleValuesForClass(
      createType<BindContext>(context),
      createType<UI5Class>(type)
    );

    expect(result).toEqual([]);
  });

  it("should handle nested extensions correctly", () => {
    const context = {
      ui5Model: {
        classes: {
          ClassA: {
            extends: {
              name: "IntermediateClass",
              kind: "class",
              extends: { name: "BaseClass", kind: "class" },
            },
          },
          ClassB: { extends: { name: "BaseClass", kind: "class" } },
        },
      },
    };

    const type = { name: "BaseClass", kind: "class" };

    const result = getPossibleValuesForClass(
      createType<BindContext>(context),
      createType<UI5Class>(type)
    );

    expect(result).toEqual(["ClassA", "ClassB"]);
  });
});

describe("buildType", () => {
  it("should build PropertyType array for UI5Any type", () => {
    const param = {
      context: createType<BindContext>({}),
      type: createType<UI5Type>({ kind: "UI5Any", name: "any" }),
      name: "testName",
      collection: false,
    };
    const result = buildType(param);
    expect(result).toEqual([
      {
        kind: "any",
        dependents: expect.any(Array),
        notAllowedElements: expect.any(Array),
        collection: false,
        reference: undefined,
      },
    ]);
  });

  it("should build PropertyType array for PrimitiveType", () => {
    const param = {
      context: createType<BindContext>({}),
      type: createType<UI5Type>({ kind: "PrimitiveType", name: "Boolean" }),
      name: "testName",
      collection: false,
    };
    const result = buildType(param);
    expect(result).toEqual([
      {
        kind: "boolean",
        dependents: expect.any(Array),
        notAllowedElements: expect.any(Array),
        possibleValue: {
          fixed: expect.any(Boolean),
          values: expect.any(Array),
        },
        collection: false,
        reference: undefined,
      },
    ]);
  });

  it("should build PropertyType array for UI5Enum type", () => {
    const param = {
      context: createType<BindContext>({}),
      type: createType<UI5Type>({
        kind: "UI5Enum",
        name: "EnumType",
        fields: [{ name: "Field1" }],
      }),
      name: "testName",
      collection: false,
    };
    const result = buildType(param);
    expect(result).toEqual([
      {
        kind: "string",
        dependents: [],
        notAllowedElements: [],
        possibleValue: {
          fixed: true,
          values: ["Field1"],
        },
        collection: false,
        reference: undefined,
      },
    ]);
  });
  it("should build PropertyType array for UI5Enum type - [FilterOperator]", () => {
    const param = {
      context: createType<BindContext>({}),
      type: createType<UI5Type>({
        kind: "UI5Enum",
        name: "FilterOperator",
        fields: [{ name: "BT" }],
      }),
      name: "testName",
      collection: false,
    };
    const result = buildType(param);
    expect(result).toEqual([
      {
        kind: "string",
        dependents: [],
        notAllowedElements: [],
        possibleValue: {
          fixed: true,
          values: ["BT"],
        },
        collection: false,
        reference: undefined,
      },
    ]);
  });
  it("should build PropertyType array for UI5Typedef type", () => {
    const param = {
      context: createType<BindContext>({}),
      type: createType<UI5Type>({
        kind: "UI5Typedef",
        name: "string",
        properties: [],
      }),
      name: "testName",
      collection: false,
    };
    const result = buildType(param);
    expect(result).toEqual([
      {
        kind: "string",
        dependents: [],
        notAllowedElements: [],
        collection: false,
        reference: undefined,
      },
    ]);
  });

  it("should build PropertyType array for UI5Class type", () => {
    const param = {
      context: createType<BindContext>({ ui5Model: {} }),
      type: createType<UI5Type>({ kind: "UI5Class", name: "ClassType" }),
      name: "testName",
      collection: false,
    };
    const result = buildType(param);
    expect(result).toEqual([
      {
        kind: expect.any(String),
        dependents: [],
        notAllowedElements: [],
        possibleElements: [],
        possibleValue: {
          fixed: false,
          values: [],
        },
        collection: false,
        reference: undefined,
      },
    ]);
  });

  it("should build PropertyType array for UnionType", () => {
    const param = {
      context: createType<BindContext>({}),
      type: createType<UI5Type>({
        kind: "UnionType",
        types: [{ kind: "PrimitiveType", name: "Boolean" }],
      }),
      name: "testName",
      collection: false,
    };
    const result = buildType(param);
    expect(result).toEqual([
      {
        kind: "boolean",
        dependents: [],
        notAllowedElements: [],
        possibleValue: {
          fixed: true,
          values: [true, false],
        },
        collection: false,
        reference: undefined,
      },
    ]);
  });
  it("should build PropertyType array for UnionType - ArrayType", () => {
    const param = {
      context: createType<BindContext>({}),
      type: createType<UI5Type>({
        kind: "UnionType",
        types: [
          {
            kind: "ArrayType",
            type: { kind: "PrimitiveType", name: "Boolean" },
          },
        ],
      }),
      name: "testName",
      collection: false,
    };
    const result = buildType(param);
    expect(result).toEqual([
      {
        kind: "boolean",
        dependents: [],
        notAllowedElements: [],
        possibleValue: {
          fixed: true,
          values: [true, false],
        },
        collection: true,
        reference: undefined,
      },
    ]);
  });

  it("should build PropertyType array for ArrayType", () => {
    const param = {
      context: createType<BindContext>({}),
      type: createType<UI5Type>({
        kind: "ArrayType",
        type: { kind: "PrimitiveType", name: "Boolean" },
      }),
      name: "testName",
      collection: false,
    };
    const result = buildType(param);
    expect(result).toEqual([
      {
        kind: "boolean",
        dependents: [],
        notAllowedElements: [],
        possibleValue: {
          fixed: true,
          values: [true, false],
        },
        collection: true,
        reference: undefined,
      },
    ]);
  });
  it("should build PropertyType array for ArrayType - undefined type", () => {
    const param = {
      context: createType<BindContext>({}),
      type: createType<UI5Type>({
        kind: "ArrayType",
      }),
      name: "testName",
      collection: false,
    };
    const result = buildType(param);
    expect(result).toEqual([]);
  });
});

describe("getConstructorParameterProperties", () => {
  const mockContext = createType<BindContext>({
    ui5Model: {
      framework: OPEN_FRAMEWORK,
    },
  });

  const mockUI5Class = createType<UI5Class>({
    kind: "UI5Class",
    abstract: false,
    extends: undefined,
    implements: [],
    ctor: undefined,
    methods: [],
    properties: [],
    fields: [],
    aggregations: [],
    associations: [],
    events: [],
    defaultAggregation: undefined,
    returnTypes: [],
  });

  it("should return an empty array when parameterProperties is empty", () => {
    const result = getConstructorParameterProperties({
      context: mockContext,
      parameterProperties: [],
      type: mockUI5Class,
    });
    expect(result).toEqual([]);
  });

  it("should return an array of BindingInfoElement when parameterProperties contains valid data", () => {
    const mockParameterProperties: UI5ConstructorParameters[] = [
      {
        kind: "UI5TypedefProp",
        name: "param1",
        type: createType<UI5Type>({ kind: "PrimitiveType", name: "Boolean" }),
        parameterProperties: [],
      },
    ];

    const result = getConstructorParameterProperties({
      context: mockContext,
      parameterProperties: mockParameterProperties,
      type: mockUI5Class,
    });

    expect(result).toEqual([
      {
        name: "param1",
        type: [
          {
            kind: "boolean",
            dependents: [],
            notAllowedElements: [],
            possibleValue: {
              fixed: true,
              values: [true, false],
            },
            collection: false,
            reference: undefined,
          },
        ],
        documentation: {
          kind: "markdown",
          value:
            "`(class) `\n\n\n\n**undefined** Boolean\n\n**undefined** undefined\n\n**undefined** undefined\n\nundefined(https://sdk.openui5.org/#/api/)",
        },
        required: true,
      },
    ]);
  });

  it("should skip parameterProperties with missing type", () => {
    const mockParameterProperties: UI5ConstructorParameters[] = [
      {
        kind: "UI5TypedefProp",
        name: "param1",
        type: undefined,
        parameterProperties: [],
      },
    ];

    const result = getConstructorParameterProperties({
      context: mockContext,
      parameterProperties: mockParameterProperties,
      type: mockUI5Class,
    });

    expect(result).toEqual([]);
  });
});

describe("getPossibleElement", () => {
  const context = createType<BindContext>({
    ui5Model: {
      framework: OPEN_FRAMEWORK,
    },
  });

  it("should return sorter possible elements when type is Sorter and parameters are not defined - fallback", () => {
    const type = createType<UI5Class>({
      name: "Sorter",
      kind: "UI5Class",
      ctor: {},
    });
    const result = getPossibleElement({ context, type });

    expect(result).toMatchSnapshot();
  });

  it("should return constructor parameter properties for sorter when name is vSorterInfo", () => {
    const type = createType<UI5Class>({
      name: "Sorter",
      kind: "UI5Class",
      ctor: {
        parameters: [
          {
            name: "vSorterInfo",
            type: "someType",
            parameterProperties: [
              {
                kind: "UI5TypedefProp",
                name: "param1",
                type: createType<UI5Type>({
                  kind: "PrimitiveType",
                  name: "Boolean",
                }),
                parameterProperties: [],
              },
            ],
          },
          {
            name: "path",
            type: "string",
            parameterProperties: [],
          },
        ],
      },
    });
    const result = getPossibleElement({ context, type });

    expect(result).toMatchSnapshot();
  });

  it("should return filter possible elements when type is Filter and vFilterInfo is not defined - fallback", () => {
    const type = createType<UI5Class>({
      name: "Filter",
      kind: "UI5Class",
      ctor: {},
    });
    const result = getPossibleElement({ context, type });

    expect(result).toMatchSnapshot();
  });

  it("should return constructor parameter properties for filter when name is vFilterInfo", () => {
    const type = createType<UI5Class>({
      name: "Filter",
      kind: "UI5Class",
      ctor: {
        parameters: [
          {
            name: "vFilterInfo",
            type: "someType",
            parameterProperties: [
              {
                kind: "UI5TypedefProp",
                name: "value1",
                type: createType<UI5Type>({
                  kind: "PrimitiveType",
                  name: "Boolean",
                }),
                parameterProperties: [],
              },
            ],
          },
          {
            name: "path",
            type: "string",
            parameterProperties: [],
          },
        ],
      },
    });

    const result = getPossibleElement({ context, type });

    expect(result).toMatchSnapshot();
  });
});
