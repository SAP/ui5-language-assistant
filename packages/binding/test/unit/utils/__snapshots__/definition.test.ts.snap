// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getPossibleElement should return constructor parameter properties for filter when name is vFilterInfo 1`] = `
Array [
  Object {
    "documentation": Object {
      "kind": "markdown",
      "value": "\`(class) Filter\`



**undefined** Boolean

**undefined** undefined

**undefined** undefined

undefined(https://sdk.openui5.org/#/api/Filter)",
    },
    "name": "value1",
    "required": true,
    "type": Array [
      Object {
        "collection": false,
        "dependents": Array [],
        "kind": "boolean",
        "notAllowedElements": Array [],
        "possibleValue": Object {
          "fixed": true,
          "values": Array [
            true,
            false,
          ],
        },
        "reference": undefined,
      },
    ],
  },
]
`;

exports[`getPossibleElement should return constructor parameter properties for sorter when name is vSorterInfo 1`] = `
Array [
  Object {
    "documentation": Object {
      "kind": "markdown",
      "value": "\`(class) Sorter\`



**undefined** Boolean

**undefined** undefined

**undefined** undefined

undefined(https://sdk.openui5.org/#/api/Sorter)",
    },
    "name": "param1",
    "required": true,
    "type": Array [
      Object {
        "collection": false,
        "dependents": Array [],
        "kind": "boolean",
        "notAllowedElements": Array [],
        "possibleValue": Object {
          "fixed": true,
          "values": Array [
            true,
            false,
          ],
        },
        "reference": undefined,
      },
    ],
  },
  Object {
    "documentation": Object {
      "kind": "markdown",
      "value": "\`(class) Sorter\`



**undefined** 

**undefined** undefined

**undefined** undefined

undefined(https://sdk.openui5.org/#/api/Sorter)",
    },
    "name": "path",
    "required": true,
    "type": Array [],
  },
]
`;

exports[`getPossibleElement should return filter possible elements when type is Filter and vFilterInfo is not defined - fallback 1`] = `
Array [
  Object {
    "documentation": Object {
      "kind": "plaintext",
      "value": "",
    },
    "name": "path",
    "type": Array [
      Object {
        "dependents": Array [],
        "kind": "string",
        "notAllowedElements": Array [],
      },
    ],
  },
  Object {
    "documentation": Object {
      "kind": "plaintext",
      "value": "",
    },
    "name": "test",
    "type": Array [
      Object {
        "dependents": Array [],
        "kind": "string",
        "notAllowedElements": Array [],
      },
    ],
  },
  Object {
    "documentation": Object {
      "kind": "plaintext",
      "value": "",
    },
    "name": "comparator",
    "type": Array [
      Object {
        "dependents": Array [],
        "kind": "string",
        "notAllowedElements": Array [],
      },
    ],
  },
  Object {
    "documentation": Object {
      "kind": "plaintext",
      "value": "",
    },
    "name": "operator",
    "type": Array [
      Object {
        "dependents": Array [],
        "kind": "string",
        "notAllowedElements": Array [],
        "possibleValue": Object {
          "fixed": true,
          "values": Array [
            "All",
            "Any",
            "BT",
            "Contains",
            "EndsWith",
            "EQ",
            "GE",
            "GT",
            "LE",
            "LT",
            "NB",
            "NE",
            "NotContains",
            "NotEndsWith",
            "NotStartsWith",
            "StartsWith",
          ],
        },
      },
    ],
  },
  Object {
    "documentation": Object {
      "kind": "plaintext",
      "value": "",
    },
    "name": "value1",
    "type": Array [
      Object {
        "dependents": Array [],
        "kind": "any",
        "notAllowedElements": Array [],
      },
    ],
  },
  Object {
    "documentation": Object {
      "kind": "plaintext",
      "value": "",
    },
    "name": "value2",
    "type": Array [
      Object {
        "dependents": Array [],
        "kind": "any",
        "notAllowedElements": Array [],
      },
    ],
  },
  Object {
    "documentation": Object {
      "kind": "plaintext",
      "value": "",
    },
    "name": "variable",
    "type": Array [
      Object {
        "dependents": Array [],
        "kind": "any",
        "notAllowedElements": Array [],
      },
    ],
  },
  Object {
    "documentation": Object {
      "kind": "plaintext",
      "value": "",
    },
    "name": "condition",
    "type": Array [
      Object {
        "dependents": Array [],
        "kind": "object",
        "notAllowedElements": Array [],
        "reference": "filters",
      },
    ],
  },
  Object {
    "documentation": Object {
      "kind": "plaintext",
      "value": "",
    },
    "name": "filters",
    "type": Array [
      Object {
        "collection": true,
        "dependents": Array [],
        "kind": "object",
        "notAllowedElements": Array [],
        "reference": "filters",
      },
    ],
  },
  Object {
    "documentation": Object {
      "kind": "plaintext",
      "value": "",
    },
    "name": "and",
    "type": Array [
      Object {
        "dependents": Array [],
        "kind": "boolean",
        "notAllowedElements": Array [],
      },
    ],
  },
  Object {
    "documentation": Object {
      "kind": "plaintext",
      "value": "",
    },
    "name": "caseSensitive",
    "type": Array [
      Object {
        "dependents": Array [],
        "kind": "boolean",
        "notAllowedElements": Array [],
      },
    ],
  },
]
`;

exports[`getPossibleElement should return sorter possible elements when type is Sorter and parameters are not defined - fallback 1`] = `
Array [
  Object {
    "documentation": Object {
      "kind": "plaintext",
      "value": "",
    },
    "name": "path",
    "required": true,
    "type": Array [
      Object {
        "dependents": Array [],
        "kind": "string",
        "notAllowedElements": Array [],
      },
    ],
  },
  Object {
    "documentation": Object {
      "kind": "plaintext",
      "value": "",
    },
    "name": "descending",
    "type": Array [
      Object {
        "dependents": Array [],
        "kind": "boolean",
        "notAllowedElements": Array [],
      },
    ],
  },
  Object {
    "documentation": Object {
      "kind": "plaintext",
      "value": "",
    },
    "name": "group",
    "type": Array [
      Object {
        "dependents": Array [],
        "kind": "string",
        "notAllowedElements": Array [],
      },
      Object {
        "dependents": Array [],
        "kind": "boolean",
        "notAllowedElements": Array [],
      },
    ],
  },
  Object {
    "documentation": Object {
      "kind": "plaintext",
      "value": "",
    },
    "name": "comparator",
    "type": Array [
      Object {
        "dependents": Array [],
        "kind": "string",
        "notAllowedElements": Array [],
      },
    ],
  },
]
`;
